# Homework Assignment 2

## Section 1: Admissions Dataset

In [2]:
#imports and data
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
admissions = pd.read_csv("../data/dataset_ucla_admissions.csv")

admissions.head()

admit    gre   gpa  prestige
0      0  380.0  3.61       3.0
1      1  660.0  3.67       3.0
2      1  800.0  4.00       1.0
3      1  640.0  3.19       4.0
4      0  520.0  2.93       4.0

**Question 1:**

    - What is the target/outcome variable? 
    - What are the features? Which features are continous or categorical? 
    - What is the null accuracy? 
    - If there are categorical variables, turn them into dummy variables and concatenate them with the data.

**Question 2:**

What do you think is the goal of using machine learning with this dataset.

**Question 3:**

Make a scatter plot of gpa vs gre color-encoded using the admit feature. What does this graph tell you?

**Question 4:**

- 1. Make a train test split using a quarter of the data for testing.
- 2. Fit a Logistic Regression model on the training data.
- 3. Evaluate the model on both the training and testing datasets using accuracy score.
- 4. Whats difference in the accuracy score of training and testing datasets? Is this evidence of overfitting?

**Question 5:** 

Use a validation curve to determine the best K-Neigbors value in a KNN model trained on this dataset and to determine the best max_depth value in a Decision Tree model as well. Which model is better? Use roc_auc as your evaluation metric.

**Question 6:**

If you work in an admissions department, would you rather have a model that minimizes false negatives or minimizes false positives? Why?

**Question 7:**

Using the Logistic Regression from Question 4, what is the probability of admission of the following prospective students:

- Student 1: GRE = 560, GPA = 3.6, Prestige = 2.
- Student 2: GRE = 750, GPA = 3.8, Prestige = 4.
- Student 3: GRE = 660, GPA = 3.1, Prestige = 1.

## Section 2: Boston Housing Data

The goal of this section will be to create a test-training split to compare multiple models on the same Boston housing data set. For example, you might choose to compare different linear models on the same data.

In this project, you'll fit two to three different models on the Boston housing data. For example, you could pick two different subsets of variables, one or more polynomial models, or any other model you'd like. 

### Here's What We Will Be Doing:

* Work with Boston housing data to predict the value of a home
* Create a test-train split of the data.
* Train your model on the training data.
* Evaluate your model on the test data.
* Repeat with other variables.
* Rank the models by how well they score on the testing data set. 
* Don't forget to interpret your findings! 




**The Scenario** 

As a real estate developer, you are being asked to model the median home price of various houses in the city of Boston using data from the US census. Your goal is to predict a continuous, numeric output (price) based on a combination of discrete features that you choose.

In [3]:
#Load the data

from sklearn.datasets import load_boston

boston = load_boston()

X = pd.DataFrame(boston.data,
                 columns=boston.feature_names)
y = pd.DataFrame(boston.target,
                 columns=['MEDV'])
#Prints the data description
print(boston['DESCR'])

.. _boston_dataset:

Boston house prices dataset
---------------------------

**Data Set Characteristics:**  

    :Number of Instances: 506 

    :Number of Attributes: 13 numeric/categorical predictive. Median Value (attribute 14) is usually the target.

    :Attribute Information (in order):
        - CRIM     per capita crime rate by town
        - ZN       proportion of residential land zoned for lots over 25,000 sq.ft.
        - INDUS    proportion of non-retail business acres per town
        - CHAS     Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)
        - NOX      nitric oxides concentration (parts per 10 million)
        - RM       average number of rooms per dwelling
        - AGE      proportion of owner-occupied units built prior to 1940
        - DIS      weighted distances to five Boston employment centres
        - RAD      index of accessibility to radial highways
        - TAX      full-value property-tax rate per $10,000
        - PTRATIO  pu

**Question 1** 

Clean up data and perform exporatory data analysis. Boston data is from scikit-learn, so it ought to be pretty clean, but you should always perform exploratory data analysis.

In [4]:
# Exploratory data analysis.

# Include: total nulls, index, data types, shape, summary statistics, and the number of unique values for each column



**Question 2**

Using `scikit-learn` Linear Regression for your model.

Pick 3-4 predictors (i.e. CRIM, ZN, etc...) that you will use to predict your target variable, MEDV.

Score and plot your predictions. What do these results tell us?

**Question 3**

Try 70/30 and 90/10 train/test splits: 70% of the data for training - 30% for testing, then 90% for training - 10% for testing.
Score and plot both of these approaches. How do your metrics change? What does this tell us about the size of training/testing splits? Include a brief written response.

**Question 4**

Try k-fold cross-validation varying the number of folds from 5 to 10.
What seems optimal? How do your scores change? Are there tradeoffs associated with choosing the number of folds? 

**Question 5**

Practice the same workflow you just worked through above, but this time use Statsmodels instead of Sklearn for linear regression. See example formulas below.

**Example: Statsmodels Formulas**

Adjust the formula using your chosen metrics from the housing data. Remember, your workflow here is the same, but the syntax is a little different. Describe your results. Are they comparable to scikit-learn's regression models?

In [5]:
# First, format your data in a DataFrame

df = pd.DataFrame(boston.data, columns=boston.feature_names)
df['MEDV'] = boston.target
df.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  MEDV  
0     15.3  396.90   4.98  24.0  
1     17.8  396.90   9.14  21.6  
2     17.8  392.83   4.03  34.7  
3     18.7  394.63   2.94  33.4  
4     18.7  396.90   5.33  36.2

In [ ]:
# Set up your new statsmodel.formula handling model
import statsmodels.formula.api as smf

# You can easily swap these out to test multiple versions/different formulas
formulas = {
    "case1": "MEDV ~ RM + LSTAT + RAD + TAX + NOX + INDUS + CRIM + ZN - 1", # - 1 = remove intercept
    "case2": "MEDV ~ NOX + RM",
    "case3": "MEDV ~ RAD + TAX"
}

model = smf.ols(formula=formulas['case1'], data=df)
result = model.fit()

result.summary()

**Question 6**

Can you optimize your R2 by selecting the best features according to pvalues? Provide your code and explain your process in a brief written response detailing whether or not this was success.

## Section 3: Mystery Dataset

In this section you will work with a "mystery" supervised classification dataset, in that you have no context to work with — the names of the features come from San Francisco neighborhoods but aren't related to anything to San Francisco.


In this section I want to devise the best possible model with the fewest number of features. Think hard about what that means. Some of the features are informative, while others are junk, your task is to find the good ones while dropping the bad ones.


I want you to test out the DT, KNN, and Logistic Regression models in order to find the best one with the best parameters.

In [6]:
mystery = pd.read_csv("../data/homework_2_data.csv", index_col=[0])

mystery.head()

haight   mission      soma    sunset  richmond  glen_park    castro  \
0 -3.126509  7.807749  0.959736 -2.248569  4.557675   1.110156  0.617296   
1  0.941629  2.935495 -0.584966 -0.880963 -1.262072   0.120989  3.970787   
2 -0.724688  4.192606 -1.139602  0.856972  0.906893  -2.799484  0.584498   
3 -2.641034  6.963189  0.962412 -0.838858  1.549620  -2.335671 -0.838585   
4  1.259116  4.741797  1.084332  0.186698  4.794145  -1.093006  2.161358   

    potrero  hayes_valley  fillmore    bernal    marina  north_beach  \
0  0.356864      0.860267 -0.395962 -1.144975 -1.805742     1.498388   
1  2.509629      1.400017  0.574335 -3.389879 -0.898020     0.512079   
2  3.468955      1.011336  0.839612 -1.384781 -0.241726     1.502255   
3  1.924333      2.794306 -0.421610 -1.273275  0.833424     2.268457   
4  3.264167     -0.060052  1.154145 -2.030773 -1.442986     0.588193   

   excelsior  target  
0   2.157730       1  
1   4.340694       1  
2  -1.539356       1  
3  -0.004678       0  
4   0.533464       1